### ISSUE TITLE
[Bug] CUDA Error: Error Code:   3 , Line  text:   initialization error

### ISSUE THREAD

**Issue by jiangxuan2023** [open-mmlab/mmdetection3d#2645]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmdetection3d/issues) and [Discussions](https://github.com/open-mmlab/mmdetection3d/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmdetection3d.readthedocs.io/en/latest/notes/faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (dev-1.x)](https://github.com/open-mmlab/mmdetection3d/tree/dev-1.x) or [latest version (dev-1.0)](https://github.com/open-mmlab/mmdetection3d/tree/dev-1.0).

### Task

I have modified the scripts/configs, or I'm working on my own tasks/models/datasets.

### Branch

1.x branch https://github.com/open-mmlab/mmdetection3d/tree/dev-1.x

### Environment

sys.platform: linux
Python: 3.8.16 (default, Jun 12 2023, 18:09:05) [GCC 11.2.0]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0,1: NVIDIA GeForce RTX 3090
CUDA_HOME: /usr/local/cuda-11.3
NVCC: Cuda compilation tools, release 11.3, V11.3.109
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.10.1

### Reproduces the problem - code sample

In det3d_dataset.py

import copy
import os
from os import path as osp
from typing import Callable, List, Optional, Set, Union
import numpy as np
import torch
from mmengine.dataset import BaseDataset
from mmengine.logging import print_log
from terminaltables import AsciiTable
from mmdet3d.registry import DATASETS
from mmdet3d.structures import get_box_type

try:
    import preprocessorcpp
except:
    print("cannot import preprocessorcpp!")

 def __init__(self,
                 data_root: Optional[str] = None,
                 ann_file: str = '',
                 metainfo: Optional[dict] = None,
                 data_prefix: dict = dict(pts='velodyne', img=''),
                 pipeline: List[Union[dict, Callable]] = [],
                 modality: dict = dict(use_lidar=True, use_camera=False),
                 default_cam_key: str = None,
                 box_type_3d: dict = 'LiDAR',
                 filter_empty_gt: bool = True,
                 test_mode: bool = False,
                 load_eval_anns: bool = True,
                 backend_args: Optional[dict] = None,
                 show_ins_var: bool = False,
                 **kwargs) -> None:
                 self.prep = preprocessorcpp.PreProcessor(config_path, data_cfg_path, training, self.rank)

Here, preprocessorcpp is a cpp/cuda object.

### Reproduces the problem - command or script

./tools/dist_train.sh CONF_PATH  2

### Reproduces the problem - error message

CUDA Error: 
    File:   /data/jiangx/code/mmlab/zhito_preprocessor_pybind11/preprocess/src/label/cnnseg_label_cuda.cu
    Line:   493
    Error Code:   3
    Line  text:   initialization error
CUDA Error: 
    File:   /data/jiangx/code/mmlab/zhito_preprocessor_pybind11/preprocess/src/label/cnnseg_label_cuda.cu
    Line:   511
    Error Code:   3
    Line  text:   initialization error
CUDA Error: 
    File:   /data/jiangx/code/mmlab/zhito_preprocessor_pybind11/preprocess/src/label/cnnseg_label_cuda.cu
    Line:   512
    Error Code:   3
    Line  text:   initialization error
CUDA Error: 
    File:   /data/jiangx/code/mmlab/zhito_preprocessor_pybind11/preprocess/src/label/cnnseg_label_cuda.cu
    Line:   513
    Error Code:   3
    Line  text:   initialization error

### Additional information

In the dataset, there is a cpp/cuda object. if num_workers = 0, it works well with single gpu or multiple gpus.  But if num_workers > 0, CUDA Error and  initialization error. How can I fix it ?
